### DISCUSSION TITLE
How to reset one particular object's state at runtime using the Python API

### DISCUSSION THREAD

**Discussion by DrawZeroPoint** [google-deepmind/mujoco#367] [Asking for Help] [answered]
Dear MuJoCo developers,

Currently, my use case involves resetting an object's pose wrt the world frame at runtime. The rationale for doing this is that the object is manipulated by a robotic manipulator, whose states should be preserved after every manipulation, whereas if we could reset only the object's states (pose to be accurate), we could avoid rerun the simulation and wait for the robot to get prepared, which is a tedious process.

In other words, this means the object can suddenly move from one pose to another across the rendered scene. I wonder if it is possible to do this using MuJoCo's Python API. A simplified code snippet I used for now looks like this:

```
tracked_object = self._data.body(self._tracked_object_name)
tracked_object.xpos = np.asarray(self._object_initial_xpos)
tracked_object.xquat = np.asarray(self._object_initial_xquat)
mujoco.mj_kinematics(self._model, self._data)
mujoco.mj_step(self._model, self._data)
```

But unfortunately, this code cannot move the object to the `initial pose` as expected, and it occasionally raises core dump at `mj_step`.

Thank you for your insights in advance, and I am willing to provide more details if necessary.

**Comment by yuvaltassa** [comment]
If I understand you correctly, the feature you are looking for is called [mocap bodies](https://mujoco.readthedocs.io/en/latest/modeling.html#motion-capture).

**Comment by DrawZeroPoint** [comment, marked as answer]
Thank you for your insight @yuvaltassa! Following the mocap bodies topic, I eventually found the solution suitable for my use case despite that it is not relevant to mocap in the end.

The "... second regular body and connect it to the mocap body with a weld equality constraint" inspired me to create a fixed body in the world and use that as an anchor of the operational body, such that by dynamically disabling the weld constraint, I could move the later with a robot, whereas by enabling that, the later will rapidly go back to the original pose defined by the anchor body.

Hereby I include some code that hopefully could be helpful for those who come across a similar problem:

The XML file:
```
  <body name="object">
    <freejoint/>
    <geom type="box" rgba="0 1 0 1"/>
  </body>
  <body name="object_anchor"/>

  <equality>
    <weld name="anchor" body1="object" body2="object_anchor"/>
  </equality>
```

The Python code:

```
self._model.equality("anchor").active = True
rospy.sleep(1.)
self._model.equality("anchor").active = False
```

Some comments about the above code:

1. The same idea can generalize to multiple objects;
2. The 'object' is the one we want to manipulate, and the 'object_anchor' defines its pose after a reset. The name 'anchor' is hard coded for now;
3. Sleep is necessary for the body to go back to the original pose. Using `rospy` is convenient in my scenario, but other options should also work. The 1 sec sleep time is an empirical value;
4. Defining the 'object_anchor' as a mocap body does not influence the results. Since in my use case I only reset the object to the fixed pose, I can omit using `mocap="true". But in other use cases like servoing, that could be a good idea.
